src/rust/engine/src/interning.rs

```diff
@@ -64,7 +62,7 @@ impl Interns {
       let key = if let Some(key) = forward_keys.get(&intern_key) {
         *key
       } else {
-        let id = self.id_generator.fetch_add(1, atomic::Ordering::SeqCst);
+        let id = self.id_generator.fetch_add(1, atomic::Ordering::Relaxed);
```

Comment:
On x86_64 the assembly is the same.

For this test rig:
```rust
$ cat rust_test/src/lib.rs 
use std::sync::atomic;

pub struct Test {
    id_generator: atomic::AtomicU64,
}

impl Test {
    #[allow(dead_code)]
    #[inline(never)]
    pub fn increment_proof_seqcst(&self) {
        let id = self.id_generator.fetch_add(1, atomic::Ordering::SeqCst);
        eprintln!("{}", id);
    }

    #[allow(dead_code)]
    #[inline(never)]
    pub fn increment_proof_relaxed(&self) {
        let id = self.id_generator.fetch_add(1, atomic::Ordering::Relaxed);
        eprintln!("{}", id);
    }
}
```

I find:
```asm
$ cargo asm rust_test::Test::increment_proof_seqcst
rust_test::Test::increment_proof_seqcst:
 sub     rsp, 72
 mov     eax, 1
 lock    xadd, qword, ptr, [rdi], rax
 mov     qword, ptr, [rsp], rax
 mov     rax, rsp
 mov     qword, ptr, [rsp, +, 8], rax
 mov     rax, qword, ptr, [rip, +, _ZN4core3fmt3num3imp52_$LT$impl$u20$core..fmt..Display$u20$for$u20$u64$GT$3fmt17h0d6ac94acfa07075E@GOTPCREL]
 mov     qword, ptr, [rsp, +, 16], rax
 lea     rax, [rip, +, .L__unnamed_1]
 mov     qword, ptr, [rsp, +, 24], rax
 mov     qword, ptr, [rsp, +, 32], 2
 mov     qword, ptr, [rsp, +, 40], 0
 lea     rax, [rsp, +, 8]
 mov     qword, ptr, [rsp, +, 56], rax
 mov     qword, ptr, [rsp, +, 64], 1
 lea     rdi, [rsp, +, 24]
 call    qword, ptr, [rip, +, _ZN3std2io5stdio7_eprint17hf8fa1c4ef4ba75d6E@GOTPCREL]
 add     rsp, 72
 ret
```

And:
```diff
$ diff <(cargo asm rust_test::Test::increment_proof_seqcst) <(cargo asm rust_test::Test::increment_proof_relaxed)
1c1
< rust_test::Test::increment_proof_seqcst:
---
> rust_test::Test::increment_proof_relaxed:
```


On arm, though, relaxed asm _should_ be different and safe and faster. I'm putting up this change though not for an unmeasured potential perf micro bump, but more to eliminate voodoo.

Reply:
And this would all be true even if the counter weren't always modified under a separate lock.

Reply:
> cargo asm

TIL. This is really neat!